File: simpledroidble/simpledroidble/src/main/java/org/simpleble/android/Adapter.kt
package org.simpleble.android

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.withContext

class Adapter private constructor(newInstanceId: Long) {
    private val _onScanStart = MutableSharedFlow<Unit>()
    private val _onScanStop = MutableSharedFlow<Unit>()
    private val _onScanUpdated = MutableSharedFlow<Peripheral>(extraBufferCapacity = 64, onBufferOverflow = BufferOverflow.DROP_OLDEST)
    private val _onScanFound = MutableSharedFlow<Peripheral>(extraBufferCapacity = 64, onBufferOverflow = BufferOverflow.DROP_OLDEST)

    private var instanceId: Long = newInstanceId
    private val callbacks = object : Callback {
        override fun onScanStart() {
            Log.d("SimpleBLE", "incomingOnScanStart")
            _onScanStart.tryEmit(Unit)
        }

        override fun onScanStop() {
            Log.d("SimpleBLE", "incomingOnScanStop")
            _onScanStop.tryEmit(Unit)
        }

        override fun onScanUpdated(peripheralId: Long) {
            _onScanUpdated.tryEmit(Peripheral(instanceId, peripheralId))
        }

        override fun onScanFound(peripheralId: Long) {
            Log.d("SimpleBLE", "incomingOnScanFound $peripheralId")
            _onScanFound.tryEmit(Peripheral(instanceId, peripheralId))
        }
    }

    init {
        nativeAdapterRegister(instanceId, callbacks)
    }

    val identifier: String get() {
        return nativeAdapterIdentifier(instanceId) ?: ""
    }

    val address: BluetoothAddress get() {
        return BluetoothAddress(nativeAdapterAddress(instanceId) ?: "")
    }

    fun scanStart() {
        nativeAdapterScanStart(instanceId)
    }

    fun scanStop() {
        nativeAdapterScanStop(instanceId)
    }

    suspend fun scanFor(timeoutMs: Int) {
        withContext(Dispatchers.IO) {
            nativeAdapterScanFor(instanceId, timeoutMs)
        }
    }

    val scanIsActive: Boolean get() {
        return nativeAdapterScanIsActive(instanceId)
    }

    fun scanGetResults(): List<Peripheral> {
        return nativeAdapterScanGetResults(instanceId).map { Peripheral(instanceId, it) }
    }

    val onScanStart get() = _onScanStart

    val onScanStop get() = _onScanStop

    val onScanUpdated get() = _onScanUpdated

    val onScanFound get() = _onScanFound

    fun getPairedPeripherals(): List<Peripheral> {
        return emptyList()
    }

    companion object {

        @JvmStatic
        fun isBluetoothEnabled(): Boolean {
            return false
        }

        @JvmStatic
        fun getAdapters(): List<Adapter> {
            val nativeAdapterIds = nativeGetAdapters()
            val adapters = ArrayList<Adapter>()

            for (nativeAdapterId in nativeAdapterIds) {
                Log.d("SimpleBLE", "Adapter found: $nativeAdapterId")
                adapters.add(Adapter(nativeAdapterId))
            }

            return adapters
        }

        @JvmStatic
        private external fun nativeGetAdapters(): LongArray
    }

    // ----------------------------------------------------------------------------

    private external fun nativeAdapterRegister(adapterId: Long, callback: Callback)

    private external fun nativeAdapterIdentifier(adapterId: Long): String?

    private external fun nativeAdapterAddress(adapterId: Long): String?

    private external fun nativeAdapterScanStart(adapterId: Long)

    private external fun nativeAdapterScanStop(adapterId: Long)

    private external fun nativeAdapterScanFor(adapterId: Long, timeout: Int)

    private external fun nativeAdapterScanIsActive(adapterId: Long): Boolean

    private external fun nativeAdapterScanGetResults(adapterId: Long) : LongArray

    // ----------------------------------------------------------------------------

    private interface Callback {
        fun onScanStart()
        fun onScanStop()
        fun onScanUpdated(peripheralId: Long)
        fun onScanFound(peripheralId: Long)
    }

}

File: simpleble/src/frontends/base/Adapter.cpp
#include <simpleble/Adapter.h>

#include "AdapterBase.h"
#include "AdapterBuilder.h"
#include "LoggingInternal.h"

using namespace SimpleBLE;

std::vector<Adapter> Adapter::get_adapters() {
    std::vector<Adapter> available_adapters;
    auto internal_adapters = AdapterBase::get_adapters();

    for (auto& internal_adapter : internal_adapters) {
        AdapterBuilder adapter(internal_adapter);
        available_adapters.push_back(adapter);
    }

    return available_adapters;
}

bool Adapter::bluetooth_enabled() { return AdapterBase::bluetooth_enabled(); }

bool Adapter::initialized() const { return internal_ != nullptr; }

void* Adapter::underlying() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().underlying();
}

std::string Adapter::identifier() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().identifier();
}

BluetoothAddress Adapter::address() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().address();
}

void Adapter::scan_start() {
    if (!initialized()) throw Exception::NotInitialized();
    if (!bluetooth_enabled()) {
        SIMPLEBLE_LOG_WARN(fmt::format("Bluetooth is not enabled."));
        return;
    }
    internal().scan_start();
}

void Adapter::scan_stop() {
    if (!initialized()) throw Exception::NotInitialized();
    if (!bluetooth_enabled()) {
        SIMPLEBLE_LOG_WARN(fmt::format("Bluetooth is not enabled."));
        return;
    }
    internal().scan_stop();
}

void Adapter::scan_for(int timeout_ms) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!bluetooth_enabled()) {
        SIMPLEBLE_LOG_WARN(fmt::format("Bluetooth is not enabled."));
        return;
    }
    internal().scan_for(timeout_ms);
}

bool Adapter::scan_is_active() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().scan_is_active();
}

std::vector<Peripheral> Adapter::scan_get_results() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().scan_get_results();
}

std::vector<Peripheral> Adapter::get_paired_peripherals() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().get_paired_peripherals();
}

void Adapter::set_callback_on_scan_start(std::function<void()> on_scan_start) {
    if (!initialized()) throw Exception::NotInitialized();

    internal().set_callback_on_scan_start(std::move(on_scan_start));
}

void Adapter::set_callback_on_scan_stop(std::function<void()> on_scan_stop) {
    if (!initialized()) throw Exception::NotInitialized();

    internal().set_callback_on_scan_stop(std::move(on_scan_stop));
}

void Adapter::set_callback_on_scan_updated(std::function<void(Peripheral)> on_scan_updated) {
    if (!initialized()) throw Exception::NotInitialized();

    internal().set_callback_on_scan_updated(std::move(on_scan_updated));
}

void Adapter::set_callback_on_scan_found(std::function<void(Peripheral)> on_scan_found) {
    if (!initialized()) throw Exception::NotInitialized();

    internal().set_callback_on_scan_found(std::move(on_scan_found));
}

AdapterBase &Adapter::internal() {
    return *internal_;
}

const AdapterBase &Adapter::internal() const {
    return *internal_;
}
